pkg/internalapis/featurestates/featurestates.go

```diff
@@ -0,0 +1,399 @@
+	CRDSingular = "cnscsisvfeaturestate"
+	// CRDPlural represent the plural name of cnscsisvfeaturestates CRD
+	CRDPlural = "cnscsisvfeaturestates"
+	// WorkLoadNamespaceLabelKey is the label key found on the workload namespace in the supervisor k8s cluster
+	WorkLoadNamespaceLabelKey = "vSphereClusterID"
+	// SVFeatureStateCRName to be used for CR in workload namespaces
+	SVFeatureStateCRName = "svfeaturestates"
+	// crUpdateRetryInterval is the interval at which all failed pending CR update/create tasks are executed
+	crUpdateRetryInterval = 1 * time.Minute
+)
+
+// pendingCRUpdates holds latest update pending to persist on the CR
+type pendingCRUpdates struct {
+	// lock to acquire before updating pendingCRUpdateTask and pendingCRCreateTask
+	lock            *sync.RWMutex
+	pendingCRUpdate map[string]*featurestatesv1alpha1.CnsCsiSvFeatureStates
+}
+
+var (
+	supervisorFeatureStatConfigMapName       string
+	supervisorFeatureStateConfigMapNamespace string
+	pendingCRUpdatesObj                      *pendingCRUpdates
+	k8sClient                                clientset.Interface
+	controllerRuntimeClient                  client.Client
+)
+
+// StartSvFSSReplicationService Starts SvFSSReplicationService
+func StartSvFSSReplicationService(ctx context.Context, svFeatureStatConfigMapName string, svFeatureStateConfigMapNamespace string) error {
+	log := logger.GetLogger(ctx)
+	log.Info("starting SvFSSReplication service")
+
+	supervisorFeatureStatConfigMapName = svFeatureStatConfigMapName
+	supervisorFeatureStateConfigMapNamespace = svFeatureStateConfigMapNamespace
+	pendingCRUpdatesObj = &pendingCRUpdates{
+		lock:            &sync.RWMutex{},
+		pendingCRUpdate: make(map[string]*featurestatesv1alpha1.CnsCsiSvFeatureStates),
+	}
+	// This is idempotent if CRD is pre-created then we continue with initialization of svFSSReplicationService
+	err := k8s.CreateCustomResourceDefinitionFromSpec(ctx, CRDName, CRDSingular, CRDPlural,
+		reflect.TypeOf(featurestatesv1alpha1.CnsCsiSvFeatureStates{}).Name(), CRDGroupName, internalapis.SchemeGroupVersion.Version, apiextensionsv1beta1.NamespaceScoped)
+	if err != nil {
+		log.Errorf("failed to create CnsCsiSvFeatureStates CRD. Error: %v", err)
+		return err
+	}
+
+	// Create the kubernetes client
+	k8sClient, err = k8s.NewClient(ctx)
+	if err != nil {
+		log.Errorf("create k8s client failed. Err: %v", err)
+		return err
+	}
+	// get kube config
+	config, err := k8s.GetKubeConfig(ctx)
+	if err != nil {
+		log.Errorf("failed to get kubeconfig. Error: %v", err)
+		return err
+	}
+	// create controller runtime client
+	controllerRuntimeClient, err = k8s.NewClientForGroup(ctx, config, CRDGroupName)
+	if err != nil {
+		log.Errorf("failed to create controllerRuntimeClient. Err: %v", err)
+		return err
+	}
+
+	// create/update feature state CRs in all workload namespaces
+	err = updateFeatureStatesCRInAllWorkLoadNamespaces(ctx, []featurestatesv1alpha1.FeatureState{})
```

Comment:
Can we not handle Add events for the SV FSS config map listener  and move the initial creation of CRs in that method?

Reply:
done listening on add config-map event.